### PR TITLE
netbsd patch: core/sys/netbsd

### PR DESCRIPTION
**Pull request by nrTQgc**
A part of NetBSD support (joakim-noah suggests to split original big pull request)

### CODE REVIEW

File: src/core/sys/netbsd/sys/cdefs.d [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * D header file for FreeBSD
+ *
```
Comment (joakim-noah) [inline src/core/sys/netbsd/sys/cdefs.d]: NetBSD header

**Comment by joakim-noah** [comment]
I diffed this PR against the pre-existing `core.sys.netbsd`: seems to mostly be a straight copy and rename, which is why the cdefs comment above is wrong and there are TODO comments that some files haven't been checked.

Fix that comment and the five remaining links to FreeBSD SVN headers, and I see no harm in merging.

**Comment by nrTQgc** [comment]
Fixed

**Comment by ibuclaw** [comment]
Do we have links to NetBSD reference C headers?

**Comment by ibuclaw** [comment]
What about Makefile changes to maybe compile and install these interface modules?

**Comment by ibuclaw** [comment]
> LDC doesn't use the upstream build system, and DMD doesn't support NetBSD...

Documentation for those who could support it when the time comes to downstream.  :-)

**Comment by nrTQgc** [comment]
> Do we have links to NetBSD reference C headers?
Actual NetBSD layout is differ. Currently I have FreeBSD layout + NetBSD changes. I should reorganize elf related files to conform NetBSD headers.

**Comment by joakim-noah** [comment]
So I guess this still has to be reorganized to conform to the NetBSD layout?  We try to keep these include files the same as the headers, so that would help.

**Comment by nrTQgc** [comment]
I added links to netbsd headers in each file and remove one file which is not related to netbsd.

File: src/core/sys/netbsd/dlfcn.d [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+     * on behavior undefined by the C standard, which might trigger a compiler
+     * diagnostic.  We intentionally declare a unique type signature to force
+     * a diagnostic should the application not cast the return value of dlfunc()
+     * appropriately.
+     */
+    struct __dlfunc_arg {
+        int     __dlfunc_dummy;
+    };
+
+    alias void function(__dlfunc_arg) dlfunc_t;
```
Comment (joakim-noah) [inline src/core/sys/netbsd/dlfcn.d]: Can you change this to the `alias =` syntax, as now done in the same FreeBSD module?  Same with the `__externC` alias below.

File: src/core/sys/netbsd/sys/link_elf.d [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    size_t          dlpi_tls_modid;
+    void*           dlpi_tls_data;
+};
+
+
+private alias extern(C) int function(dl_phdr_info*, size_t, void *) dl_iterate_phdr_cb;
+private alias extern(C) int function(dl_phdr_info*, size_t, void *) @nogc dl_iterate_phdr_cb_ngc;
+extern int dl_iterate_phdr(dl_iterate_phdr_cb __callback, void*__data);
+extern int dl_iterate_phdr(dl_iterate_phdr_cb_ngc __callback, void*__data) @nogc;
+
```
Comment (joakim-noah) [inline src/core/sys/netbsd/sys/link_elf.d]: Remove blank line.

**Comment by nrTQgc** [comment]
I fixed issues (alias & empty line). I did not find anything new in current FreeBSD sources. I agree to merge this PR now. I will wait while LDC team merge current druntime master to their repository and after this I can refresh unit tests status (new LDC, new druntime, and new NetBSD 7.1 too) and may be send additional PRs.

File: src/core/sys/netbsd/dlfcn.d [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * http://cvsweb.netbsd.org/bsdweb.cgi/~checkout~/src/include/dlfcn.h
+ */
+module core.sys.netbsd.dlfcn;
+
+public import core.sys.posix.dlfcn;
+
+version (NetBSD):
+extern (C):
+nothrow:
```
Comment (joakim-noah) [inline src/core/sys/netbsd/dlfcn.d]: FreeBSD file has `@nogc` attribute here, missing in yours.  Why not add that attribute to all the header files?

**Comment by joakim-noah** [comment]
btw, it will take a while before these patches get to ldc, so let me know if you'd rather work with the beta 1.2 branch or master 1.3 branch and I will cherry-pick these commits for ldc earlier.

**Comment by joakim-noah** [comment]
Pinging @klickverbot, maybe Kai can announce NetBSD support at DConf. ;)

**Comment by nrTQgc** [comment]
>joakim-noah 
I am not sure about ldc branches 1.2 vs 1.3. I suppose 1.3 branch is better - the probaility of merge conflicts is lower

**Comment by joakim-noah** [comment]
I've cherry-picked the three NetBSD druntime commits into [the master 1.3 branch](https://github.com/ldc-developers/druntime/commits/ldc).  You will need to check out the latest druntime HEAD manually, until the submodule is updated.

**Pull request merged**